Task 7

IvanPavlov90 opened the pull request.

IvanPavlov90 committed:
create methods for task 7.1, will think about  necessity of one array. Maybe it's useless.

IvanPavlov90 committed:
comments for functions

IvanPavlov90 committed:
create solution for task 7.2, case with start with negative number isn't included

IvanPavlov90 committed:
finish task 7.2

IvanPavlov90 committed:
add string check and rename functions 7.1

IvanPavlov90 committed:
add methods that throw err if user expression contains symbols or repeat separators

IvanPavlov90 committed:
fix error message in task 7.1

IvanPavlov90 committed:
add some mthods for task 7.3

IvanPavlov90 committed:
Add solution for task 7.3

IvanPavlov90 committed:
fix error treatment at 7.3

IvanPavlov90 committed:
fix incapsulation and inputs:

- add input to the task 7.1;
- fix incapsulation in task 7.3 by creating module;
- fix error treatment in 7.2 and 7.1 tasks;

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+        return deleteDuplicates(textArray, repeats);
+    } catch (e) {
+        throw e;
+    }
```
Why catch and rethrow the error?

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+        userInput.addEventListener("blur", (event) => {
+            let result = changeText (event.target.value);
+            output.innerHTML = result;
+        });
```
`blur` isn't really a good idea - not intuitive. I expected the result to update on every keystroke, but it didn't work even on enter + there's no button. I thought that the script hasn't loaded

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+        checkText(userText);
+        let word = [];
+        let repeats = [];
+        let textArray = userText.split("");
```
`Array.from(userText)` is a better and more straightforward solution

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+function comparisingWithSeparators (letter) {
+    const separators = [" ", ",", "?", "!", ".", ":", ";"];
+    return separators.includes(letter);
+}
```
You should be more careful with function names, since they don't even have signatures. They can take whatever and return whatever. Name should be crystal clear, like `isSeparator(letter)`

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+        let word = [];
+        let repeats = [];
+        let textArray = userText.split("");
+        for (let i = 0; i < textArray.length; i++) {
```
...though you can just iterate through a string just like in C# with `for..of`

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+    try {
+        checkText(userText);
+        let word = [];
+        let repeats = [];
```
`repeatedLetters`

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+    } else {
+        return true;
+    }
+}
```
This function either returns true, either undefined, which is falsy, but still shady. Additionally, change the name to something clearer like `addLetterToWordIfNotRepeated` (reflects what this function does) or refactor the code to divide adding a letter from repetition checks.

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+                    if (!repeats.includes(textArray[i].toLowerCase())) {
+                        repeats.push(textArray[i].toLowerCase());
+                    }
+                }
```
You can combine these `if`s, or

Here you can create a function like `addIfNotExist` instead of moving `findRepetitiveSymbolsInWord`'s logic where it is now. I think this way the code will read better. 
By the way, there's `Set` object, which acts like `HashSet` is C# - ensures it doesn't contain duplicates.

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+/* This function compares symbols from the sentence with separators. 
+   Returns true if symbol is separator. */
+   
+function comparisingWithSeparators (letter) {
```
If you want Intellisense to help you with docs, you can use jsdocs - such comments start with a slash and 2 asterisks (`/** */`)

STeeL835 commented on Task 7/Task 7.1/script.js:
```diff
@@ -0,0 +1,75 @@
+   Then it deletes such symbols from user's text. */ 
+
+function deleteDuplicates (textArray, repeats) {
+    for (let i = 0; i < textArray.length; i++) {
+        if (repeats.includes(textArray[i].toLowerCase())) {
+            textArray.splice(i, 1);
+            i = i - 1;
+        }
+    }
```
Or, you could go from beginning and add only those characters which are not in `repeats`.

Also you can simplify it by using `filter` function on the array

STeeL835 commented on Task 7/Task 7.2/script.js:
```diff
@@ -0,0 +1,115 @@
+    try {
+        let userInput = document.getElementById("userExpression");
+        let output = document.getElementById("result");
+        userInput.addEventListener("blur", (event) => {
+            let result = calculator (event.target.value);
+            output.innerHTML = "Your result is " + result;
+        });
+    } catch (e) {
+        alert(e.message);
+    }
```
You register a callback which will be called when the event is fired. The registering part does and waits nothing, because it doesn't call the callback, so catching an error will do nothing here, unless there's an error during setup. To actually catch errors that happen during event processing, you should catch errors there.

STeeL835 commented on Task 7/Task 7.2/script.js:
```diff
@@ -0,0 +1,115 @@
+        let digitsArray = getDigits(userExpression);
+        let result = getResult (digitsArray, ariphmeticSigns);
+        return result;
+    } catch (e) {
+        throw e;
+    }
```
Why catch and throw errors without additional logic?

STeeL835 commented on Task 7/Task 7.2/script.js:
```diff
@@ -0,0 +1,115 @@
+
+/* This function checks if format of user's expression is correct or not. */
+
+function checkExpression (expression) {
+    expression = cutRightPartOfExpression(expression.trim());
+    if (expression.startsWith("*") || expression.startsWith("/")) {
+        throw new Error ("Your expression isn't valid.");
+    } else {
+        const regexp = /[0-9]/;
+        const symbolsRegexp = /[A-Za-zА-Яа-яЁё]/;
+        let index = expression.search(regexp);
+        let symbolsIndex = expression.search(symbolsRegexp);
+        if (index == -1 || index > 1 || symbolsIndex != -1) {
+            throw new Error ("Your expression isn't valid.");
+        }
+        return expression;
+    }
+}
```
This all can be achieved with one regexp

STeeL835 commented on Task 7/Task 7.2/script.js:
```diff
@@ -0,0 +1,115 @@
+function getAriphmeticSigns(expression) {
+    let signs = [];
+    const separators = ["+", "-", "/", "*"];
+    for (let i = 0; i < expression.length; i ++) {
+        if (separators.includes(expression[i])) {
+            signs.push(expression[i]);
+        }
+    }
```
```js
expression.filter(char => separators.includes(char))
``` 
(if expression is an Array) is better than `for`. You can either convert string to an array with `Array.from` or steal `filter` from Array prototype and call it on a string with `<function>.call()` (but that is an advanced technique, you'll learn it later)

STeeL835 commented on Task 7/Task 7.2/script.js:
```diff
@@ -0,0 +1,115 @@
+
+/* This method cuts everything that is located after sign "=" */
+
+function cutRightPartOfExpression (expression) {
+    let index = expression.indexOf("=");
+    let result = expression.slice(0, index + 1);
+    return result;
+}
```
You can combine this and checks with a regexp (to get expression without `=` you can use groups)

STeeL835 commented on Task 7/Task 7.2/script.js:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+/* This method provides protection on cases when there are no digits 
+   between separators. Like 4/5++2= */
+
+function checkDigitsArray(digitsArray) {
+    const regexp = /[0-9]/;
+    for (let i = 0; i < digitsArray.length; i++) {
+        let index = digitsArray[i].search(regexp);
+        if (index == -1) {
+            throw new Error ("Your expression isn't valid.");
+        }
+    }
+}
```
You won't need this with a complex RegExp from a comment above

STeeL835 commented on Task 7/Task 7.3/storage.js:
```diff
@@ -0,0 +1,82 @@
+    };
+
+    checkIncomingElement (obj) {
+        if (typeof obj !== "object" || Array.isArray(obj)) {
```
Another way is 
```js
!(obj instanceof Array)
```
but you can also receive a RegExp object for example. I guess you can use prototype information to allow only plain objects
```js
Object.getPrototypeOf(obj) === Object.prototype
```
but prototypes are also an advanced thing

STeeL835 commented on Task 7/Task 7.3/storage.js:
```diff
@@ -0,0 +1,82 @@
+        this._storage[objIndex] = obj;
+    };
+
+    getObjectID(obj) {
+        obj.id = this._idCount.toString();
+        this._idCount++;
+    };
```
Don't call a function "get" if it doesn't gets (returns) anything. "set" is better alternative

STeeL835 commented on Task 7/Task 7.3/storage.js:
```diff
@@ -0,0 +1,82 @@
+    getById(id) {
+        this.checkTypeOfID(id);
+        let resultObj = this._storage.find(item => item.id === id);
+        if (resultObj === undefined) {
+            return null;
```
better leave it as undefined. It's not a C#, there is a difference between `null` and `undefined`, and it is quite literal. If an element doesn't exist, it is _not defined_, if it doesn't exist but explicitly (a place is reserved, but no object) - it's `null`

IvanPavlov90 replied:
There was special condition in the task, that if object is not found by id, you should return null. Maybe it was meant "null"???

STeeL835 commented on Task 7/Task 7.3/storage.js:
```diff
@@ -0,0 +1,82 @@
+        }
+    };
+
+    getAll() {
+        this._storage.forEach((element) => {
+            console.log(element);
+        })
+    };
```
get should return something. I think you meant `logAll`

STeeL835 commented on Task 7/Task 7.3/storage.js:
```diff
@@ -0,0 +1,82 @@
+        };
+        let count = 0;
+        let resultObj = this._storage.find(item => item.id === id);
+        for (let prop in resultObj) {
+            if (prop !== "id") {
+                resultObj[prop] = data[count];
+                count++;
+            }
+        }
```
Also, there's an `Object.assign()` and spread syntax

STeeL835 commented on Task 7/Task 7.3/storage.js:
```diff
@@ -0,0 +1,82 @@
+    replaceById(id, obj) {
+        this.checkTypeOfID(id);
+        this.checkIncomingElement(obj);
+        let resultObj = this._storage.find(item => item.id === id);
+        obj.id = id;
+        let objIndex = this._storage.indexOf(resultObj);
+        this._storage[objIndex] = obj;
```
You can use `findIndex`

IvanPavlov90 committed:
fixes for 7.1

1. get rid of extra error throw;
2. change "blur" event to click on special btn;
3. use Array.from insteead of split;
4. change name of comaparsingWithSeparators function;

IvanPavlov90 committed:
fix for 7.1

1. add special function for searching duplicated letters in word, this function replaces two if's;
2. renamed set objects;

IvanPavlov90 committed:
use jsdocs notation and change method for creating final result

IvanPavlov90 committed:
fix task 7.2

1. add special button and change event type
2. fix error throw
3. change cutRightPartOfExpression function

IvanPavlov90 committed:
change replaceById and updateById methods in task 7.3

IvanPavlov90 committed:
change getAll method task 7.3

IvanPavlov90 committed:
fix regexp for checking user expression

IvanPavlov90 committed:
little fix for 7.3

IvanPavlov90 committed:
fix for 7.2
